service/kernel/oneapi/select_indexed.h

```diff
@@ -33,35 +33,140 @@ namespace internal
 {
 namespace selection
 {
-
-class QuickSelectIndexed
+class SelectIndexed
 {
 public:
     struct Result
     {
         UniversalBuffer values;
         UniversalBuffer indices;
 
-        Result(ExecutionContextIface& context, uint32_t K,  uint32_t nVectors, TypeId valueType, TypeId indexType, services::Status* status)
-            : values(context.allocate(valueType, nVectors * K, status)),
-              indices(context.allocate(indexType, nVectors * K, status))
+        Result() {}
+        Result(ExecutionContextIface & context, uint32_t K, uint32_t nVectors, TypeId valueType, services::Status * status)
+            : values(context.allocate(valueType, nVectors * K, status)), indices(context.allocate(TypeIds::id<int>(), nVectors * K, status))
         {}
     };
+    struct Params
+    {
+        Params(uint32_t K, TypeId fptype, uint32_t size, daal::algorithms::engines::EnginePtr eng) : nK(K), type(fptype), dataSize(size), engine(eng)
+        {}
+        uint32_t nK;
+        TypeId type;
+        uint32_t dataSize;
+        daal::algorithms::engines::EnginePtr engine;
+    };
+
+public:
+    virtual ~SelectIndexed() {}
+    virtual Result select(const UniversalBuffer & dataVectors, uint32_t K, uint32_t nVectors, uint32_t vectorSize, uint32_t lastVectorSize,
+                          uint32_t vectorOffset, services::Status * status)                    = 0;
+    virtual Result & select(const UniversalBuffer & dataVectors, uint32_t K, uint32_t nVectors, uint32_t vectorSize, uint32_t lastVectorSize,
+                            uint32_t vectorOffset, Result & result, services::Status * status) = 0;
+    static void convert(const UniversalBuffer & indices, const UniversalBuffer & labels, uint32_t nVectors, uint32_t vectorSize,
+                        uint32_t vectorOffset, services::Status * status);
+    void selectLabels(const UniversalBuffer & distances, const UniversalBuffer & dataLabels, uint32_t nK, uint32_t nVectors, uint32_t vectorSize,
+                      uint32_t vectorOffset, uint32_t labelOffset, Result & result, services::Status * status)
+    {
+        select(distances, nK, nVectors, vectorSize, vectorSize, vectorOffset, result, status);
```

Comment:
It seems like `select` and `convert` function implement **template method** pattern. Do we need to declare them `public`?